rebase -i: support --ignore-date

r1walz opened the pull request:
Let's review this one.

r1walz commented on t/t3431-rebase-options-compatibility.sh:
```diff
@@ -81,7 +81,15 @@ test_expect_success '--committer-date-is-author-date works with interactive back
 	sed -ne "/^author /s/.*> //p" head >authortime &&
 	sed -ne "/^committer /s/.*> //p" head >committertime &&
-	test_cmp authortime committertime
+	test_cmp authortime committertime &&
```
Ignore test file.

tgummerer commented on sequencer.c:
```diff
@@ -1005,6 +1006,17 @@ static int run_git_commit(struct repository *r,
+			if (opts->ignore_date) {
+				struct strbuf new_author = STRBUF_INIT;
+
+				while (*author != '>')
```
We should check that we don't run over the end of the string here, e.g. if somebody changes the author script on disk, we still shouldn't read some random memory.

Even better I think we can replace this `while` loop with a `memchr` to get the length of the bits we need to split off, and, add that to the `strbuf` using `strbuf_add` and then `datestamp` it.

Other than being easier to read, I think this also allows us to plug a memory leak, as we are leaking the original `author` right now.  (I think we're also leaking it currently, but now we'd leak two instances)

I also wonder if it would be better to traverse this string from the end, rather than from the front.  I don't think we allow `>` in the author name, so this should work fine in all cases, but I'd still feel more comfortable going over something that isn't user controlled.

Also I think `author` could be `NULL` here right, if the `is_rebase_i` clause is not satisfied.  We should also check for that I think.

r1walz replied:
> Also I think author could be NULL here right, if the is_rebase_i clause is not satisfied. We should also check for that I think.

`--ignore-date` is only set by rebase -i so I think we are clear on this one.

tgummerer replied:
> `--ignore-date` is only set by rebase -i so I think we are clear on this one.

Ah yes of course, that makes sense.  It would be nice to have a `BUG()` statement then, indicating that we know about this, and helping developers in the future if they want to add another codepath here, e.g. if they want to start having a similar functionality in `cherry-pick`.

tgummerer commented on sequencer.c:
```diff
@@ -3481,6 +3513,9 @@ static int do_merge(struct repository *r,
 		}
+		if (opts->ignore_date)
+			argv_array_pushf(&cmd.env_array,
+					 "GIT_AUTHOR_DATE=%ld", time(NULL));
```
Convenient :)

r1walz replied:
AFAIK, --date sets AUTHOR TZ to UTC. But we are using local TZ in the logic. Should be changed ASAP.

tgummerer commented on sequencer.c:
```diff
@@ -1497,6 +1511,17 @@ static int try_to_commit(struct repository *r,
+		datestamp(&new_author);
+
+		author = strbuf_detach(&new_author, NULL);
+	}
```
This is the same code as above, should we factor this out into a separate function?

newren replied:
...and perhaps also move the opts->committer_date_is_author_date block into that new function since it is also repeated?

tgummerer reviewed (commented):
Thanks!  Quite a lot to unpack in that single line there, though I think it's relatively easy to fix :)

Nice work, looking forward to reading the docs and the tests for this as well.

newren commented on builtin/rebase.c:
```diff
@@ -473,6 +475,8 @@ int cmd_rebase__interactive(int argc, const char **argv, const char *prefix)
 			 &opts.committer_date_is_author_date,
 			 N_("make committer date match author date")),
+		OPT_BOOL(0, "ignore-date", &opts.ignore_date,
+			 N_("ignore author date and use current date")),
```
I know we're just copying git-am and earlier git-rebase here, but the name "--ignore-date" is just not descriptive enough.  In am's case it's slightly more defensible as an email message has a single date and it's known that it is used to set the author date.  But even there it feels too vague.  In the context of rebase, it's just bad -- there are clearly two dates we are working with.

We really ought to rename this to something else ("--author-date-is-committer-date"? "--ignore-author-date"?), and just keep --ignore-date around as an alias for the more precise name.

newren reviewed (commented):
Only two small comments other than what Thomas found.

r1walz committed:
rebase -i: support --ignore-date

rebase am already has this flag to "lie" about the author date
by changing it to the committer (current) date. Let's add the same
for interactive machinery.

Signed-off-by: Rohit Ashiwal <rohit.ashiwal265@gmail.com>

tgummerer commented on Documentation/git-rebase.txt:
```diff
@@ -388,8 +388,9 @@ See also INCOMPATIBLE OPTIONS below.
 --ignore-date::
-	This flag is passed to 'git am' to change the author date
-	of the rebased commits (see linkgit:git-am[1]).
+--author-date-is-committer-date::
```
This is now not quite correct for am based rebases.  `--author-date-is-committer-date` doesn't exist there, and I don't see this patch implementing it.  Maybe introducing that flag in a preparatory patch would be nice.

tgummerer commented on Documentation/git-rebase.txt:
```diff
@@ -388,8 +388,9 @@ See also INCOMPATIBLE OPTIONS below.
-	of the rebased commits (see linkgit:git-am[1]).
+--author-date-is-committer-date::
+	Lie about the author date by re-setting it to the value
+	same as comitter (current) date. This implies --force-rebase.
```
s/comitter/committer/

tgummerer commented on sequencer.c:
```diff
@@ -1507,6 +1528,13 @@ static int try_to_commit(struct repository *r,
+		ignore_author_date(&author);
+		if (flags & AMEND_MSG) {
+			free(amend_author);
+			amend_author = (char *)author;
```
It seems to me that `amend_author` is only used to free `author` when we get it from `get_author`, instead of from the parameter that's passed in (and should not be free'd.

So it seems to me that we should always assign `author` to `amend_author` here to avoid ano memory leaks.  At that point `amend_author` is no longer the right name, so maybe we should rename it to `author_to_free` or something in a preparatory patch?

r1walz replied:
Yes, it is indeed used to do the aforementioned. Let's account all your suggested changes in some commits.

tgummerer commented on t/t3433-rebase-options-compatibility.sh:
```diff
@@ -84,4 +84,22 @@ test_expect_success '--committer-date-is-author-date works with interactive back
 	test_cmp authortime committertime
 '
 
+# Checking for +0000 in author time is enough since default
```
That doesn't sound quite right.  The default time zone when committing should always be the timezone that's set up where the tests are run.  If we alwayse get UTC in the tests here, that's probably and indication that the code doesn't deal with timezones correctly.

And we don't set the timezone to UTC in the tests anywhere, as far as I can see.

r1walz replied:
> The default time zone when committing should always be the timezone that's set up where the tests are run.

This is correct. But ...

> And we don't set the timezone to UTC in the tests anywhere, as far as I can see.

... we're actually setting timezone in [`test-lib.sh`](https://github.com/git/git/blob/v2.22.0/t/test-lib.sh#L368).

tgummerer replied:
> ... we're actually setting timezone in test-lib.sh.

Ah fair enough, I missed that. Thanks for the pointer.  I tried to `cat-file` a commit in t1010 to test this, but only now realized that we're using `test_tick` there, which confusingly sets the timezone to -0700.    So we're all good here, sorry for the noise.

tgummerer commented on sequencer.c:
```diff
@@ -1013,10 +1025,17 @@ static int run_git_commit(struct repository *r,
+			if (opts->ignore_date) {
+				if (!author)
+					BUG("rebase -i must set author "
+					    "before committing tree");
```
Maybe expand this to something like "ignore_date can only be used with rebase -i, which must set the author befor committing the tree"?

r1walz commented on t/t3433-rebase-options-compatibility.sh:
```diff
@@ -84,4 +84,22 @@ test_expect_success '--committer-date-is-author-date works with interactive back
+test_expect_success '--ignore-date works with am backend' '
+	git commit --amend --date="$GIT_AUTHOR_DATE" &&
+	git rebase --ignore-date HEAD^ &&
+	git cat-file commit HEAD | sed -e "/^\$/q" >head &&
```
How about using `git show HEAD --pretty="format:%ai"` instead?

tgummerer replied:
Yeah, I think that would be preferrable.

r1walz committed:
sequencer: rename amend_author to author_to_free

The purpose of amend_author was to free() the malloc()'d string
obtained from get_author() when AMEND_MSG flag was set. But now,  we
also use it to free() the new parsed author so, the name is no longer
relevant. Rename it to something meaningful.

Signed-off-by: Rohit Ashiwal <rohit.ashiwal265@gmail.com>

r1walz committed:
rebase: add --author-date-is-committer-date

Commit 554377d ("rebase -i: support --ignore-date", 2019-08-05)
introduced --ignore-date flag to interactive rebase, but the name
is actually very vague in context of rebase -i since there are two
dates we can work with. Add an alias to convey the precise purpose.

Signed-off-by: Rohit Ashiwal <rohit.ashiwal265@gmail.com>

r1walz commented:
@tgummerer @newren Let's have a final call before sending this to the ML.

newren commented on sequencer.c:
```diff
@@ -1425,7 +1447,7 @@ static int try_to_commit(struct repository *r,
 		return -1;
 	if (!(flags & AMEND_MSG) && opts->committer_date_is_author_date &&
-	    setenv_committer_date_to_author_date())
+	    !opts->ignore_date && setenv_committer_date_to_author_date())
```
What if someone passes both --committer-date-is-author-date and --ignore-date?  Do they expect author date and committer date to get swapped?  Right now, this appears to make --ignore-date override --committer-date-is-author-date -- perhaps we should error out instead?  (Or maybe this is fine as-is because git-am also surprisingly just makes --ignore-date silently override --committer-date-is-author-date, without documenting it, so you've at least matched the behavior elsewhere...)

Not sure you need to do anything here, especially since you do match git-am behavior, but it might be worth pointing out in your cover letter.

newren reviewed (commented):
Looks good, but (as a very micro nit:) you have a double space after a comma in the commit message; might be nice to clean that up.

newren reviewed (approved):
Wahoo!  I like it.

tgummerer reviewed (commented):
> Commit 554377d ("rebase -i: support --ignore-date", 2019-08-05)

Unfortunately you can't refer to a commit like that when it's part of a patch series, as the commit ID will be different once Junio actually applies the earlier patch.  Instead, the patch can just be referred as "a previous patch" or something like that.

I would also make what you have as the second commit in the series ("sequencer: rename amend_author to author_to_free") the first commit, as it's really a preparatory step that makes the commit after it easier to read.  Right now it's more of an "oops, we should have named this differently, let's do it now" commit, which we generally prefer not to do in this project.

Other than those two nits the code looks good to me 🎉

r1walz commented:
--- sent to the [Mailing List](https://public-inbox.org/git/20190806173638.17510-1-rohit.ashiwal265@gmail.com/) ---
